packages/mcp/src/server.ts

```diff
@@ -105,91 +107,97 @@ export class MCPServer extends MCPServerBase {
-          name: tool.name,
-          description: tool.description,
-          inputSchema: tool.parameters.jsonSchema,
-        })),
-      };
-    });
+    if (!this.listToolsHandlerIsRegistered) {
+      this.listToolsHandlerIsRegistered = true;
+      this.server.setRequestHandler(ListToolsRequestSchema, async () => {
+        this.logger.debug('Handling ListTools request');
+        return {
+          tools: Object.values(this.convertedTools).map(tool => ({
+            name: tool.name,
+            description: tool.description,
+            inputSchema: tool.parameters.jsonSchema,
+          })),
+        };
+      });
+    }
   }
 
   /**
    * Register the CallTool handler for executing a tool by name.
    */
   private registerCallToolHandler() {
-    this.server.setRequestHandler(CallToolRequestSchema, async request => {
-      const startTime = Date.now();
-      try {
-        const tool = this.convertedTools[request.params.name];
-        if (!tool) {
-          this.logger.warn(`CallTool: Unknown tool '${request.params.name}' requested.`);
-          return {
-            content: [{ type: 'text', text: `Unknown tool: ${request.params.name}` }],
-            isError: true,
-          };
-        }
-
-        this.logger.debug(`CallTool: Invoking '${request.params.name}' with arguments:`, request.params.arguments);
-
-        const validation = tool.parameters.validate?.(request.params.arguments ?? {});
-        if (validation && !validation.success) {
-          this.logger.warn(`CallTool: Invalid tool arguments for '${request.params.name}'`, {
-            errors: validation.error,
-          });
-          return {
-            content: [{ type: 'text', text: `Invalid tool arguments: ${JSON.stringify(validation.error)}` }],
-            isError: true,
-          };
-        }
-        if (!tool.execute) {
-          this.logger.warn(`CallTool: Tool '${request.params.name}' does not have an execute function.`);
-          return {
-            content: [{ type: 'text', text: `Tool '${request.params.name}' does not have an execute function.` }],
-            isError: true,
-          };
-        }
-
-        const result = await tool.execute(validation?.value, { messages: [], toolCallId: '' });
-        const duration = Date.now() - startTime;
-        this.logger.info(`Tool '${request.params.name}' executed successfully in ${duration}ms.`);
-        return {
-          content: [
-            {
-              type: 'text',
-              text: JSON.stringify(result),
-            },
-          ],
-          isError: false,
-        };
-      } catch (error) {
-        const duration = Date.now() - startTime;
-        if (error instanceof z.ZodError) {
-          this.logger.warn('Invalid tool arguments', {
-            tool: request.params.name,
-            errors: error.errors,
-            duration: `${duration}ms`,
-          });
+    if (!this.callToolHandlerIsRegistered) {
```

Comment:
If you flip this the diff will be a lot smaller! Easier to review:
`if (this.callToolHandlerIsRegistered) return`